lib/shelley/src/Cardano/Wallet/Shelley/MinimumUTxO.hs

```diff
@@ -0,0 +1,101 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE TypeApplications #-}
+
+-- |
+-- Copyright: © 2020 IOHK
+-- License: Apache-2.0
+--
+-- Computing minimum UTxO values.
+--
+module Cardano.Wallet.Shelley.MinimumUTxO
+    ( computeMinimumCoinForUTxO
+    ) where
+
+import Prelude
+
+import Cardano.Wallet.Primitive.Types.Address
+    ( Address (..) )
+import Cardano.Wallet.Primitive.Types.Coin
+    ( Coin (..) )
+import Cardano.Wallet.Primitive.Types.MinimumUTxO
+    ( MinimumUTxO (..) )
+import Cardano.Wallet.Primitive.Types.TokenBundle
+    ( TokenBundle (..) )
+import Cardano.Wallet.Primitive.Types.TokenMap
+    ( TokenMap )
+import Cardano.Wallet.Primitive.Types.Tx
+    ( TxOut (..) )
+import Cardano.Wallet.Shelley.Compatibility
+    ( toCardanoTxOut )
+import Data.IntCast
+    ( intCast, intCastMaybe )
+import Data.Word
+    ( Word64 )
+import GHC.Stack
+    ( HasCallStack )
+import Numeric.Natural
+    ( Natural )
+
+import qualified Cardano.Api as Cardano
+import qualified Data.ByteString as BS
+
+computeMinimumCoinForUTxO :: HasCallStack => MinimumUTxO -> TokenMap -> Coin
+computeMinimumCoinForUTxO = \case
+    MinimumUTxONone ->
+        const (Coin 0)
+    MinimumUTxOConstant c ->
+        const c
+    MinimumUTxOForShelleyBasedEra era pp -> \m ->
+        let result = Cardano.calculateMinimumUTxO era
+                (tokenMapToTxOut era m)
+                (Cardano.fromLedgerPParams era pp)
+        in
+        case result of
+            Left e -> error $ unwords
+                [ "computeMinimumCoinForUTxO:"
+                , "unexpected error:"
+                , show e
+                ]
+            Right value ->
+                unsafeValueToWalletCoin value
+  where
+    tokenMapToTxOut
+        :: Cardano.ShelleyBasedEra era
+        -> TokenMap
+        -> Cardano.TxOut Cardano.CtxTx era
+    tokenMapToTxOut era m =
+        toCardanoTxOut era $ TxOut dummyAddress $ TokenBundle dummyCoin m
+      where
+        -- TODO:
+        -- Use a suitable length that is at least as long as any valid address:
+        dummyAddress :: Address
+        dummyAddress = Address $ BS.pack $ 97 : replicate 28 0
```

Comment:
Reminder: replace these magic numbers (`97` and `28`) with named identifiers and/or add explanations for their provenance.